contributors/design-proposals/testing/trigger-pr-tests.md

```diff
@@ -0,0 +1,92 @@
+# Triggering PR tests
+
+## Objective
+
+This document describes a major change to the way the `trigger` plugin determines if test jobs should be started on a pull request (PR).
+
+## Current situation
+
+PR test jobs are started by the trigger plugin on **trusted PR** events, or when a untrusted PR becomes trusted.
+> A PR is considered trusted if the author is a member of the *trusted organization* for the repository or if such a member has left an `/ok-to-test` command on the PR.
+
+You can easily spot an untrusted PR by its `needs-ok-to-test` label.
+
+## Problems
+
+### Reviews are not parsed by the trigger plugin
+
+We only parse PR comments for the `/ok-to-test` command, not review comments.
+
+@spiffxp
+> I still see people using /ok-to-test in PR reviews (as comments, aka no approval or change requested) and nothing happening
+
+@cjwagner
+> We aren't currently fetching the reviews, just the comments.
+
+### API tokens consumption
+
+@cjwagner
+> The `trigger` plugin has to list all comments to see if an `/ok-to-test` command exists every time it wants to check that a non-member PR is trusted. Checking reviews and review comments for this command would require at least 2 additional API tokens every time.
+
+> Fetching the reviews and review comments would use a lot more API tokens. If we switch to a positive label we don't need to check anything except the labels though.
+
+### Code complexity
+
+The `trigger` code is already quite complex, and also parsing the reviews would make it worse.
+
+@cjwagner
+> We could check `review_comment` events and remove the `do-not-merge/needs-ok-to-test` label if the comment contains `/ok-to-test`, but we would still need to list all reviews and review comments when checking if a non-member PR is trusted to prevent race conditions with other events.
+
+> Using a positive label instead of a negative one avoids all the race conditions that force us to list comments and check for `/ok-to-test` instead of just checking for a label.
+
+## Proposal
+
+We suggest introducing a new label named `ok-to-test` that would be required on any non-member PR before automatic test jobs can be started by the `trigger` plugin.
+
+This label could only be added by members of the *trusted organization* for the repository.
+
+We will also deprecate the `/ok-to-test` command.
```

Comment:
We will still need this command. This is how org members will add the `ok-to-test` label to authorize non-member PRs.

Reply:
Ok, my bad, I thought we wanted to stop parsing the comments (and still not parse reviews) to save the API tokens.

Reply:
Indeed the [way](https://help.github.com/articles/applying-labels-to-issues-and-pull-requests/) to add labels by hand isn't really practical... I thought it was available directly in the issue/PR page.

Reply:
yeah, there's a way in the page but it performs a set on the labels from the browser, not an "add to whatever currently exists" so if the labels change while you're editing them (IE prow labels something) and then you finish it will set the labels you see, and remove any you didn't have locally 🙄

Reply:
> Ok, my bad, I thought we wanted to stop parsing the comments (and still not parse reviews) to save the API tokens.

We will save a lot of API tokens by avoiding listing the comments, reviews, and review comments every time we need to check if a PR is trusted.
Instead we can handle the corresponding events (issue_comment, pull_request_review, and pull_request_review_comment) with a single GenericCommentEvent handler that adds the `ok-to-test` label in response to `/ok-to-test` from an org-member. From then on whenever we need to check if the PR is trusted, we just see if it is authored by an org member or has the `ok-to-test` label (no need to list everything).